File: backend/pkg/model/entity/canvas/course.go
package canvas

const (
	StudentEnrollment  = "StudentEnrollment"
	TaEnrollment       = "TaEnrollment"
	TeacherEnrollment  = "TeacherEnrollment"
	ObserverEnrollment = "ObserverEnrollment"
	DesignerEnrollment = "DesignerEnrollment"
)

type Enrollment struct {
	EnrollmentType  string `json:"type,omitempty"`
	Role            string `json:"role,omitempty"`
	RoleID          int64  `json:"role_id,omitempty"`
	UserID          int64  `json:"user_id,omitempty"`
	EnrollmentState string `json:"enrollment_state,omitempty"`
}

type Term struct {
	ID            int64   `json:"id,omitempty"`
	Name          string  `json:"name,omitempty"`
	StartAt       *string `json:"start_at,omitempty"`
	EndAt         *string `json:"end_at,omitempty"`
	CreatedAt     *string `json:"created_at,omitempty"`
	WorkflowState string  `json:"workflow_state,omitempty"`
}

type Teacher struct {
	ID          int64  `json:"id,omitempty"`
	DisplayName string `json:"display_name,omitempty"`
}

type Course struct {
	ID                     int64         `json:"id,omitempty"`
	UUID                   string        `json:"uuid,omitempty"`
	Name                   string        `json:"name,omitempty"`
	CourseCode             string        `json:"course_code,omitempty"`
	Enrollments            []*Enrollment `json:"enrollments,omitempty"`
	Teachers               []*Teacher    `json:"teachers,omitempty"`
	AccessRestrictedByDate *bool         `json:"access_restricted_by_date,omitempty"`
	Term                   *Term         `json:"term,omitempty"`
}

type Assignment struct {
	ID                      int64       `json:"id"`
	Description             *string     `json:"description,omitempty"`
	DueAt                   *string     `json:"due_at,omitempty"`
	UnlockAt                *string     `json:"unlock_at,omitempty"`
	LockAt                  *string     `json:"lock_at,omitempty"`
	PointsPossible          *float64    `json:"points_possible,omitempty"`
	CourseID                int64       `json:"course_id"`
	Name                    string      `json:"name"`
	NeedsGradingCount       *int32      `json:"needs_grading_count,omitempty"`
	HtmlURL                 string      `json:"html_url"`
	SubmissionTypes         []string    `json:"submission_types"`
	AllowedExtensions       []string    `json:"allowed_extensions,omitempty"`
	HasSubmittedSubmissions bool        `json:"has_submitted_submissions"`
	Published               bool        `json:"published"`
	SubmissionsDownloadURL  string      `json:"submissions_download_url,omitempty"`
	Submission              *Submission `json:"submission,omitempty"`
}

type Submission struct {
	ID           int64         `json:"id"`
	SubmittedAt  *string       `json:"submitted_at,omitempty"`
	Grade        *string       `json:"grade,omitempty"`
	AssignmentID int64         `json:"assignment_id"`
	UserID       int64         `json:"user_id"`
	Late         bool          `json:"late"`
	Attachments  []*Attachment `json:"attachments"`
}

type Attachment struct {
	ID          int64  `json:"id"`
	UUID        string `json:"uuid"`
	FolderID    *int64 `json:"folder_id,omitempty"`
	DisplayName string `json:"display_name"`
	Filename    string `json:"filename"`
	URL         string `json:"url,omitempty"`
	Size        int64  `json:"size"`
	Locked      bool   `json:"locked"`
	MimeClass   string `json:"mime_class,omitempty"`
	ContentType string `json:"content-type,omitempty"`
}

File: frontend/src/model/canvas/course.ts
import { User } from "./user";

export interface Course {
    id: number;
    uuid: string;
    name: string;
    course_code: string;
    enrollments: Enrollment[];
    teachers: Teacher[];
    term: Term;
}

export interface Teacher {
    id: number;
    display_name: string;
}

interface Term {
    id: number;
    name: string;
    start_at?: string | null;
    end_at?: string | null;
    created_at?: string | null;
    workflow_state: string;
}

export type EnrollmentRole = "TaEnrollment" | "StudentEnrollment" | "TeacherEnrollment" | "DesignerEnrollment" | "ObserverEnrollment";

export interface Enrollment {
    type: string;
    role: EnrollmentRole;
    role_id: number;
    user_id: number;
    enrollment_state: string;
}


export interface Assignment {
    id: number;
    key: number;
    needs_grading_count: number | null;
    description: string | null;
    due_at?: string | null;
    unlock_at?: string;
    lock_at?: string;
    points_possible?: number;
    course_id: number;
    name: string;
    html_url: string;
    submission_types: string[];
    allowed_extensions: string[];
    published: boolean;
    has_submitted_submissions: boolean;
    submission?: Submission;
}

export interface AssignmentDate {
    id: number;
    base: boolean;
    title: string;
    due_at: string | null;
    unlock_at: string | null;
    lock_at: string | null;
}

export interface AssignmentOverride {
    id: number;
    pub_id: number;
    assignment_id: number;
    quiz_id: number;
    context_module_id: number;
    student_ids: number[];
    group_id: number;
    course_section_id: number;
    title: string;
    due_at: string | null;
    all_day: boolean;
    all_day_date: string;
    unlock_at: string | null;
    lock_at: string | null;
}

export type WorkflowState = "submitted" | "unsubmitted" | "graded" | "pending_review";

export interface Submission {
    id: number;
    key: number;
    grade: string | null;
    submitted_at?: string;
    assignment_id: number;
    user_id: number;
    late: boolean;
    attachments?: Attachment[];
    workflow_state: WorkflowState;
}

export interface Attachment {
    key: React.Key;
    user?: User;
    user_id: number;
    submitted_at?: string;
    grade: string | null;
    id: number;
    late: boolean;
    uuid: string;
    folder_id: number | null;
    display_name: string;
    filename: string;
    "content-type": string;
    url: string;
    size: number;
    locked: boolean;
    mime_class: string;
    preview_url: string;
}